[CIVIC-1105] Add Merlin UI migration connector.

richardgaunt opened the pull request:
https://salsadigital.atlassian.net/browse/CIVIC-1105

## Checklist before requesting a review

- [x] I have formatted the subject to include ticket number as `[CIVIC-123] Verb in past tense with dot at the end.`
- [x] I have added a link to the JIRA ticket
- [x] I have provided information in `Changed` section about WHY something was done if this was not a normal implementation
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have run new and existing relevant tests locally with my changes, and they passed
- [x] I have provided screenshots, where applicable

## Changed
1. Added configuration form to allow upload of local merlin ui extracted content JSON or remote endpoint retrieval of files
2. Added configuration file for CivicTheme migration group
3. Added migration stub file for CivicTheme civic_page migration
4. Added process to update the CivicTheme migration for the extracted_content json files
5. Added tests to test end to end the migration process and form functionality
6. Split business logic into separate methods
7. Split source JSON validation into a service
8. Added kernel test to test the service

I did not add behat tests testing the file upload widget as not sure how to get this to work with behat. 

## Screenshots

https://user-images.githubusercontent.com/57734756/195228214-10da0c24-99ce-48ed-86d0-c65aae0b9441.mp4

richardgaunt commented on docroot/modules/custom/civictheme_migration/assets/migrate_plus.migration.civictheme_page.yml:
```diff
@@ -0,0 +1,54 @@
+  # Under 'ids', we identify source fields populated above which will uniquely
+  # identify each imported item. The 'type' makes sure the migration map table
+  # uses the proper schema type for stored the IDs.
+  ids:
```
Every migration needs an ID field - I have named the field ID and this is how it should be configured in Merlin UI for all migrations.

richardgaunt commented on docroot/modules/custom/civictheme_migration/assets/migrate_plus.migration.civictheme_page.yml:
```diff
@@ -0,0 +1,54 @@
+    - name: id
+      label: "ID"
+      selector: id
+    - name: title
```
Mapping to the extracted content from Merlin.

richardgaunt commented on docroot/modules/custom/civictheme_migration/src/Form/CivicThemeMigrationConfigurationForm.php:
```diff
@@ -0,0 +1,500 @@
+    }
+    $migration_config_file_name = $this->extensionPathResolver->getPath('module', 'civictheme_migration') . '/assets/migrate_plus.migration.civictheme_page.yml';
+    $migration_config = file_get_contents($migration_config_file_name);
+    $migration_config = $this->yaml->decode($migration_config);
```
Didn't need to do any scrappy `file_get_contents` and `str_replace` - migration files are YML so we can load, decode, update and encode.

AlexSkrypnyk commented on docroot/modules/custom/civictheme_migrate/src/Form/CivicThemeMigrationConfigurationForm.php:
```diff
@@ -0,0 +1,500 @@
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    $config = $this->config('civictheme_migrate.settings');
+    if ($this->isRetrieveFilesSubmit($form_state)) {
+      $urls = explode("\r\n", $form_state->getValue('endpoint'));
```
windows

```
$urls = explode("\n",  str_replace("\r\n", "\n", $form_state->getValue('endpoint')));
```

AlexSkrypnyk replied:
please move URL retrieval this into a public method. The submit handler only should be preparing data, passing to other methods and producing success/messages. All business logic should be happening elsewhere.

AlexSkrypnyk commented on docroot/modules/custom/civictheme_migrate/src/Form/CivicThemeMigrationConfigurationForm.php:
```diff
@@ -0,0 +1,500 @@
+    }
+
+    if ($this->isSaveConfigurationSubmit($form_state) || $this->isGenerateMigrationSubmit($form_state)) {
+      $config->set('configuration_files', $form_state->getValue('configuration_files'));
```
same - separate public method please

AlexSkrypnyk commented on docroot/modules/custom/civictheme_migrate/src/Form/CivicThemeMigrationConfigurationForm.php:
```diff
@@ -0,0 +1,500 @@
+    if ($this->isSaveConfigurationSubmit($form_state) || $this->isGenerateMigrationSubmit($form_state)) {
+      $config->set('configuration_files', $form_state->getValue('configuration_files'));
+    }
+    $config->set('migration_type', $form_state->getValue('migration_type'));
```
this code is too confusing for me. parts of it execute things, parts are saving configurations.

richardgaunt replied:
Split into methods to be more explicit on what is happening.

AlexSkrypnyk commented on docroot/modules/custom/civictheme_migrate/src/Form/CivicThemeMigrationConfigurationForm.php:
```diff
@@ -0,0 +1,500 @@
+   *   Formstate object.
+   */
+  protected function generateMigrationConfigurationSubmit(array &$form, FormStateInterface $form_state) {
+    $file_ids = $form_state->getValue('configuration_files');
```
can all of this be moved to a MigrationManager class and handled there. The config form class (this class) is only responsible for rendering form and dispatching submissions

AlexSkrypnyk commented on scripts/custom/drupal-install-site-4-enable-migration.sh:
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
```
Please add enabling of this module after `echo "  > Enable development module."` in scripts/custom/drupal-install-site-3...` and remove this file.

AlexSkrypnyk commented on tests/behat/features/migration.feature:
```diff
@@ -0,0 +1,75 @@
+@civictheme @civictheme_migrate
```
please rename this file to `migrate.feature`

AlexSkrypnyk commented on tests/behat/fixtures/merlin_ui_extracted_content.json:
```diff
@@ -0,0 +1,22 @@
+[
```
please remove `merlin ui` from the title

Please name this `civictheme_migrate.page_content_1.json`

AlexSkrypnyk commented on tests/behat/fixtures/merlin_ui_extracted_content2.json:
```diff
@@ -0,0 +1,22 @@
+[
```
`civictheme_migrate.page_content_2.json`

AlexSkrypnyk commented on tests/behat/fixtures/merlin_ui_extracted_content_invalid.json:
```diff
@@ -0,0 +1 @@
+Invalid JSON file.
```
👍

AlexSkrypnyk reviewed (changes requested):
@richardgaunt 
Please see comments

The biggest items are:
1. Still needs validation of fields
2. Business logic in the form submit - should be moved to a class.

richardgaunt commented on docroot/modules/custom/civictheme_migrate/assets/migrate_plus.migration.civictheme_page.yml:
```diff
@@ -0,0 +1,54 @@
+  # are the source field names we want to populate (the names to be used as
+  # sources in the process configuration below). For each field we're importing,
+  # we provide a label (optional - this is for display in migration tools) and
+  # an xpath for retrieving that value. It's important to note that this xpath
```
Updated comment - copied form example migration file for explanation of setup.

richardgaunt commented on composer.json:
```diff
@@ -20,6 +20,7 @@
         "drush/drush": "^11",
         "govcms/govcms": "^2.18",
         "oomphinc/composer-installers-extender": "^2.0",
+        "opis/json-schema": "^2.3",
```
JSON Validation library - uses JSON Schema `json-schema.org/` which will work well for our purposes.

richardgaunt commented on docroot/modules/custom/civictheme_migrate/civictheme_migrate.module:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  /** @var \Drupal\civictheme_migrate\CivicThemeMigrateValidator $schema_validator */
+  $schema_validator = \Drupal::service('civictheme_migrate.validator');
```
Validation of source JSON file added here.

richardgaunt commented:
@AlexSkrypnyk Ready for re-review.

1. Added validation of JSON schema for source JSON files - which can be built upon with migration enhancement with just changes to the schema file - happy to give demo explanation of it. This is enclosed in a service.
2.  Separated validation from config saving in form submit to make it less confusing.
3. Updated naming from `civictheme_migration` to `civictheme_migrate`

AlexSkrypnyk commented on docroot/modules/custom/civictheme_migrate/tests/src/Kernel/CivicThemeMigrateValidatorTest.php:
```diff
@@ -0,0 +1,67 @@
+    $json = json_decode(file_get_contents(__DIR__ .
+      '/../../fixtures/migration_source_json/' . $file_path));
+    $validation_result = $this->validator->validate($json, $schema_id);
+    $this->assertEquals($expected, $validation_result->isValid());
```
the code in the form does not use `isValid()` and uses `validate()` which produces certain messages based on the type of an error.

Can we please add the tests for `validate()` and error retrieval

AlexSkrypnyk commented on tests/behat/features/migrate.feature:
```diff
@@ -0,0 +1,76 @@
+    And I visit "[TEST] Migrated Content 10"
+
+  @api
+  Scenario Outline: Only administrator can access the CivicTheme migration configuration form
```
nit: please move this to the top

richardgaunt commented on composer.d10.json:
```diff
@@ -24,6 +24,8 @@
         "drupal/layout_builder_restrictions": "@dev",
         "drupal/linkit": "@dev",
         "drupal/menu_block": "dev-8.x-1.8",
+        "drupal/migrate_plus": "^6.0",
```
I have moved them to the module but i have preserved here as I could not get the build working on CI (local worked ok) with using a path dependency to the new module. I note that no other module has depedencies in their composer file and not here as well.

richardgaunt commented on docroot/modules/custom/civictheme_migrate/civictheme_migrate.module:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Checks that the file is recognized as a valid JSON file.
+ *
+ * @param \Drupal\file\FileInterface $file
```
Updated

richardgaunt commented on docroot/modules/custom/civictheme_migrate/civictheme_migrate.module:
```diff
@@ -0,0 +1,43 @@
+ */
+function civictheme_migrate_validate_json(FileInterface $file, $schema_id) {
+  $errors = [];
+  $json = json_decode(file_get_contents($file->getFileUri()));
```
Moved all validation to validator

richardgaunt commented on docroot/modules/custom/civictheme_migrate/src/Form/CivicThemeMigrateConfigurationForm.php:
```diff
@@ -0,0 +1,534 @@
+   *   Formstate object.
+   */
+  protected function retrieveRemoteFilesSubmit(array &$form, FormStateInterface $form_state) {
+    $urls = explode("\n", str_replace("\r\n", "\n", $form_state->getValue('endpoint')));
```
All moved to Manager class - but kept config saving in form as I think that better aligns with purpose of form.

richardgaunt commented on composer.d10.json:
```diff
@@ -34,8 +36,10 @@
         "vlucas/phpdotenv": "^5.1",
-        "webflo/drupal-finder": "^1.2"
+        "webflo/drupal-finder": "^1.2",
+        "wikimedia/composer-merge-plugin": "^2.0"
```
Added `composer-merge-plugin` to D9 and D10 composer files and completed a `composer update --lock` to implement the merge.

AlexSkrypnyk replied:
@richardgaunt 
the packages above are still not removed - can you please remove

richardgaunt replied:
Done - building now on CI.
